### ISSUE TITLE
xtypes is not working in our environment

### ISSUE THREAD

**Issue by javiersorianoruiz** [eclipse-cyclonedds/cyclonedds-cxx#329]
Hello everybody.

We (@Bit-Seq, @jordialexalar and I) have been trying examples with **XTypes and cyclonedds-cxx** and they don't work.

It works under Python (with cyclonedds-python).

But with cyclonedds-cxx it doesn't work for us. We don't know if we are doing something wrong.

We work with this IDL (I have a publisher and subscriber sending and receiving this message. We are using the helloword example provided by the same cyclonedds-cxx as base of our testing):

```
module HelloWorldData
{
  @mutable
  @autoid(HASH)
  struct Msg
  {
    @key long userID;
    string message;
  };
};
```

We have created a new publisher and subscriber with the previous message but modified (new variable `long extra_value;`)

```
module HelloWorldData
{
  @mutable
  @autoid(HASH)
  struct Msg
  {
    @key long userID;
    long extra_value;
    string message;
  };
};

```

They don't speak between them. 

That's to say, subscriber1 (original message) only receives messages from publisher1, and it doesn't receive any message from the publisher2 (modified message).

On the other hand, subscriber2 only receives messages from publisher2, and it doesn't receive any message from the publisher1.

We have checked with `cmake -LAH` both parameters that are indicated in the link [What syntax and datatypes are supported in IDL](https://cyclonedds.io/content/blog/xtypes-features.html):

```
ENABLE_TOPIC_DISCOVERY:BOOL=ON
ENABLE_TYPE_DISCOVERY:BOOL=ON
```

So, we think that everything is properly configured.

Are we doing something wrong in cyclonedds-cxx with XTypes or is there an actual problem with XTypes in C++?, because with cyclonedds-python a similar example worked well.

All the binaries, (publisher1,subscriber1, publisher2 and subscriber2) are launched on 4 running containers over the same docker image. So, configurations and settings are exactly the same for all the binaries.

The Dockerfile to reproduce the problem is this:

```
FROM        ubuntu:20.04

# update and install dependencies
RUN         apt-get update \
                && apt-get install -y \
                    software-properties-common \
                    wget \
                && add-apt-repository -y ppa:ubuntu-toolchain-r/test \
                && apt-get update \
                && apt-get install -y \
                    make \
                    git \
                    curl \
                    vim \
                && apt-get install -y cmake \
                && apt-get install -y gcc-11 \
                && apt-get install -y g++-11 \
                && update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-11 100 \
                && update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-11 100

WORKDIR /root/

#cyclonedds compilation and installation
RUN git clone https://github.com/eclipse-cyclonedds/cyclonedds.git --branch 0.10.1
WORKDIR /root/cyclonedds/
RUN mkdir build
WORKDIR /root/cyclonedds/build
RUN cmake -DCMAKE_INSTALL_PREFIX=/usr/local/lib/cyclonedds -DBUILD_EXAMPLES=ON ..
RUN cmake --build .
RUN cmake --build . --target install

#googletest compilation and installation
WORKDIR /root/
RUN git clone https://github.com/google/googletest.git -b release-1.12.1
WORKDIR /root/googletest/
RUN mkdir build
WORKDIR /root/googletest/build
RUN cmake ..
RUN make
RUN make install



#cyclonedds-cxx compilation and installation
WORKDIR /root/
RUN git clone https://github.com/eclipse-cyclonedds/cyclonedds-cxx.git --branch 0.10.1
WORKDIR /root/cyclonedds-cxx/
RUN mkdir build
WORKDIR /root/cyclonedds-cxx/build
RUN cmake   -DCMAKE_INSTALL_PREFIX=/usr/local/lib/cyclonedds-cxx \
            -DCMAKE_PREFIX_PATH=/usr/local/lib/cyclonedds \
            -DBUILD_EXAMPLES=ON \
            -DCMAKE_BUILD_TYPE=Debug \
            ..
RUN cmake --build .
RUN cmake --build . --target install


```

Kind regards.

**Comment by javiersorianoruiz** [comment]
We have just tested in a docker with gcc10 and enabling the cmake flags recommended in the link [What syntax and datatypes are supported in IDL](https://cyclonedds.io/content/blog/xtypes-features.html) and it continues not working.

cyclonedds compilation

```
cmake -DCMAKE_INSTALL_PREFIX=/usr/local/lib/cyclonedds \
            -DBUILD_EXAMPLES=ON \
            -DENABLE_TOPIC_DISCOVERY=ON \
            -DENABLE_TYPE_DISCOVERY=ON \
            ..

```
cyclonedds-cxx compilation

```
cmake   -DCMAKE_INSTALL_PREFIX=/usr/local/lib/cyclonedds-cxx \
            -DCMAKE_PREFIX_PATH=/usr/local/lib/cyclonedds \
            -DBUILD_EXAMPLES=ON \
            -DCMAKE_BUILD_TYPE=Debug \
            -DENABLE_TOPIC_DISCOVERY=ON \
            -DENABLE_TYPE_DISCOVERY=ON \
            ..
```

**Comment by reicheratwork** [comment]
@javiersorianoruiz I will take a look at this problem today

**Comment by reicheratwork** [comment]
@javiersorianoruiz Can replicate this issue

Interestingly enough, a C++ writer CAN talk to a C reader but a C writer CANNOT talk to a C++ reader.

Taking a deeper look :)

**Comment by javiersorianoruiz** [comment]
OK, perfect, @reicheratwork let me know if you get something

**Comment by reicheratwork** [comment]
@javiersorianoruiz Very close to a solution, I will send you a request for a review of the PR sometime tomorrow or friday, okay?

**Comment by javiersorianoruiz** [comment]
Ok, perfect

**Comment by javiersorianoruiz** [comment]
Hello @reicheratwork 

I have downloaded your fork and branch reicheratwork:fix_issue_329 and by the moment I have found this problem.

I have started to test appendable topics.

I have a publisher1 and subscriber1 which sends and receives this message:

```
module module_test
{
  @topic
  @appendable
  struct Msg
  {
    long userID;
    string message;
  };
}; 
```

I have another publisher2 and subscriber2 which sends and receives this message (notice the new field `int32 extra_value;`)

```
module module_test
{
  @topic
  @appendable
  struct Msg
  {
    long userID;
    string message;
    int32 extra_value;
  };
};
```

In all the cases XTypes is working except in the case publisher 1 is sending and subscriber2 is receiving, this error is shown by the subscriber2:

```
./subscriber2
=== [Subscriber] Create reader.
=== [Subscriber] Wait for message.
1669115723.680887 [0]     recvUC: data(application, vendor 1.16): 110e52e:bec3b69c:63e4b1f4:203 #1: deserialization module_test_Msg/module_test::Msg failed (for reasons unknown)
```

Note: subscriber 2 is properly receiving messages from publisher2

**Comment by reicheratwork** [comment]
Fixed this new issue, and also rebased the commits onto the current master, as this is required to have CycloneDDS-CXX compile with the current master of CycloneDDS-C

**Comment by javiersorianoruiz** [comment]
Hello @reicheratwork 

I don't know if I am doing something wrong. I have installed both versions (CycloneDDS and CycloneDDS-CXX) using the current master, and it continues not working, but now the problem is another.

Now: 

- publisher1 sends and only subscriber1 receives (subscriber2 doesn't receive).
- publisher2 sends and only subscriber2 receives (subscriber1 doesn't receive).

You can see all the test for cpp that I am doing in this repository (see subfolder `cpp`)

https://github.com/javiersorianoruiz/cyclonedds-xtypes-testing

If you need more information about these tests, please let me know, maybe there is something wrong in them, but if not, then the XTypes in cpp it is still not working.

**Comment by reicheratwork** [comment]
I don't see the issue by just looking through it.

Also, by implementing your models and publisher/subscriber programs verbatim, I cannot seem to reproduce your issue.
The publisher sending the "expanded" message gets it delivered to both the "basic" and "expanded" subscribers, and the same for the "basic" publisher.

Can you describe how you start the different processes and in which order?

**Comment by javiersorianoruiz** [comment]
Yes, of course.

I have 4 different binaries:

- ./appendable/pub_sub1/appendable_publisher1
- ./appendable/pub_sub1/appendable_subscriber1
- ./appendable/pub_sub2/appendable_publisher2
- ./appendable/pub_sub2/appendable_subscriber2

When I launch appendable_publisher1 and appendable_subscriber1 I see how the subscriber is receiving the message:

```
./appendable/pub_sub1/appendable_subscriber1 
=== [Subscriber] Create reader.
=== [Subscriber] Wait for message.
=== [Subscriber] Message received:
    userID  : 1
    Message : "Hello World"
=== [Subscriber] Done.

```

But when I launch appendable_publisher1 and appendable_subscriber2 I see how the subscriber is not receiving the message:

```
./appendable/pub_sub2/appendable_subscriber2 
=== [Subscriber] Create reader.
=== [Subscriber] Wait for message.

```

On the other hand,  when I launch appendable_publisher2 and appendable_subscriber2 I see how the subscriber is receiving the message:

```
./appendable/pub_sub2/appendable_subscriber2
=== [Subscriber] Create reader.
=== [Subscriber] Wait for message.
=== [Subscriber] Message received:
    userID  : 111
    Message : "Hello World"
    extra_value : "222"
=== [Subscriber] Done.

```

But when I launch appendable_publisher2 and appendable_subscriber1 I see how the subscriber is not receiving the message:

```
./appendable/pub_sub1/appendable_subscriber1
=== [Subscriber] Create reader.
=== [Subscriber] Wait for message.

```

I hope that I have explained enough well.

**Comment by reicheratwork** [comment]
Hmm, this is really weird...

And you don't get any read errors or the like anymore?

**Comment by javiersorianoruiz** [comment]
No, I don't get any read errors, the subscriber remains without receiving anything, waiting for a message in order to show its content.

**Comment by javiersorianoruiz** [comment]
I have tried the different order that you indicated, and the results are the same. 

When I first start appendable_subscriber2 and appendable_subscriber1, and only then start appendable_publisher1 -> Only appendable_subscriber1 receives the message

When I first start appendable_subscriber2 and appendable_subscriber1, and only then start appendable_publisher2 -> Only appendable_subscriber2 receives the message

I am working with `gcc/g++ (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0`

I have also compiled cyclonedds with these cmake parameters:

```
cmake -DCMAKE_INSTALL_PREFIX=/usr/local/lib/cyclonedds \
     -DBUILD_EXAMPLES=ON -DCMAKE_BUILD_TYPE=Debug -DENABLE_TOPIC_DISCOVERY=ON \
     - DENABLE_TYPE_DISCOVERY=ON .. 

```

I have also compiled cyclonedds-cxx with these cmake parameters:

```
cmake   -DCMAKE_INSTALL_PREFIX=/usr/local/lib/cyclonedds-cxx \
            -DCMAKE_PREFIX_PATH=/usr/local/lib/cyclonedds \
            -DBUILD_EXAMPLES=ON \
            -DCMAKE_BUILD_TYPE=Debug \
            -DENABLE_TOPIC_DISCOVERY=ON \
            -DENABLE_TYPE_DISCOVERY=ON \
            ..
```

I don't know if you can work with docker. In my GitHub project for XTypes testing, there is a Dockerfile that can be easily modified to compile the current master branch of cyclonedds and cyclonedds-cxx and this way testing exactly with the same environment as me.

**Comment by reicheratwork** [comment]
As I am not running the programs under docker and not seeing any issues, could you try to run these programs in just a command line interface, not inside docker, and see what you get, that way we can eliminate one variable?

Also, you can have logging enabled so that CycloneDDS spits out log files.
You can do this by creating an xml file with the following contents:
```
<CycloneDDS>
<Domain>
	<Tracing>
		<Verbosity>finest</Verbosity>
		<OutputFile>cdds.log.${CYCLONEDDS_PID}</OutputFile>
	</Tracing>
</Domain>
</CycloneDDS>
```
and then setting the environment variable ``CYCLONEDDS_URI`` to the location of the config file you want to provide it.

Maybe this will give us some clues as to what is going on here?

**Comment by javiersorianoruiz** [comment]
Ok, I have launched appendable_subscriber2 and appendable_subscriber1, and only then start appendable_publisher1 then I have 3 files generated:

- cdds.log.223  
- cdds.log.558
- cdds.log.236


How can I send you these files?

**Comment by reicheratwork** [comment]
> Ok, I have launched appendable_subscriber2 and appendable_subscriber1, and only then start appendable_publisher1 then I have 3 files generated:
> 
> * cdds.log.223
> * cdds.log.558
> * cdds.log.236
> 
> How can I send you these files?

You can put them in your ``cyclonedds-xtypes-testing`` repository.
Though it's probably better if you rename the files to something more descriptive.

**Comment by javiersorianoruiz** [comment]
I have just downloaded compile and installed cyclonedds (C and CXX, both versions) current master with the last commit that you indicated, and the results are the same, subscriber2 doesn't receive from publisher1. 

I don't know what happens in my environment, because you are telling me that my examples are working in your environment, aren't they?

How can I send you the logs files recorded in the last testing?

**Comment by reicheratwork** [comment]
> One other check: Which commits of CycloneDDS-C and CycloneDDS-CXX are you using to compile? I am currenly using the following: CycloneDDS-C: f28f6f97137aa7758430150b4437c773ca7a021e CycloneDDS-CXX: [372217a](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/commit/372217a4726216e79b0c868f4fc4bc9070160d32)

Seems like there were quite a few commits to both CyclonedDDS-C added yesterday, so I have to rebase the PR and fix some changes to filenames/defines.
So using the current master of CycloneDDS-C will not let the current status of the PR build succesfully until i have fixed this.

**Comment by javiersorianoruiz** [comment]
Sorry I didn't see your comment about the log files

I have put them in my cyclonedds-xtypes-testing repository (dds_logs_files folder in the root directory)

**Comment by reicheratwork** [comment]
> I have just downloaded compile and installed cyclonedds (C and CXX, both versions) current master with the last commit that you indicated, and the results are the same, subscriber2 doesn't receive from publisher1.
> 
> I don't know what happens in my environment, because you are telling me that my examples are working in your environment, aren't they?
> 
> How can I send you the logs files recorded in the last testing?

So you are using the current ``master`` branch of both ``CycloneDDS-C`` and ``CycloneDDS-CXX`` from the ``eclipse`` repositories?

The fix for this issue is still not in eclipse:cyclonedds-cxx:master, so it might be that you are compiling against something else entirely.

I am working with the following code:
[github.com/eclipse-cyclonedds/cyclonedds:f28f6f97137aa7758430150b4437c773ca7a021e ](https://github.com/eclipse-cyclonedds/cyclonedds/commit/f28f6f97137aa7758430150b4437c773ca7a021e)
and
[github.com/reicheratwork/cyclonedds-cxx/commit:372217a4726216e79b0c868f4fc4bc9070160d32](https://github.com/reicheratwork/cyclonedds-cxx/commit/372217a4726216e79b0c868f4fc4bc9070160d32)

Note that I am referring to not the current most recent version of master of CDDS-C, and a specific commit on CDDS-CXX on my personal repo (as the changes have not yet been merged to eclipse CDDS-CXX)

**Comment by reicheratwork** [comment]
> Sorry I didn't see your comment about the log files
> 
> I have put them in my cyclonedds-xtypes-testing repository (dds_logs_files folder in the root directory)
> 
> On the other hand, do I have to await your new versions of cyclonedds C and CXX ? Or can I do any other test to get more information about what is happening?

I will have a look at them.
But please check which commits of the CycloneDDS C and C++ code you are using, so that we can be sure that we are talking about the same thing here :)

**Comment by javiersorianoruiz** [comment]
I am not sure if I am doing it rightly, because I can now compile my XTypes tests.

For cyclonedds I have done

```
git clone git@github.com:eclipse-cyclonedds/cyclonedds.git
cd cyclonedds
git reset --hard f28f6f97137aa7758430150b4437c773ca7a021e
cmake -DCMAKE_INSTALL_PREFIX=/usr/local/lib/cyclonedds -DBUILD_EXAMPLES=ON -DCMAKE_BUILD_TYPE=Debug -DENABLE_TOPIC_DISCOVERY=ON -DENABLE_TYPE_DISCOVERY=ON ..
cmake --build .
cmake --build . --target install
```

For cyclonedds-cxx I have done

```
git clone git@github.com:reicheratwork/cyclonedds-cxx.git -b fix_issue_329
cd cyclonedds-cxx
git reset --hard 372217a4726216e79b0c868f4fc4bc9070160d32
cmake -DCMAKE_INSTALL_PREFIX=/usr/local/lib/cyclonedds-cxx -DCMAKE_PREFIX_PATH=/usr/local/lib/cyclonedds -DBUILD_EXAMPLES=ON -DCMAKE_BUILD_TYPE=Debug -DENABLE_TOPIC_DISCOVERY=ON -DENABLE_TYPE_DISCOVERY=ON ..
cmake --build .

This error appears

[ 92%] Linking CXX executable ../../bin/ddscxxHelloworldSubscriber
/usr/bin/ld: ../../lib/libddscxx.so.0.11.0: undefined reference to `ddsi_default_plist_participant'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/helloworld/CMakeFiles/ddscxxHelloworldSubscriber.dir/build.make:109: bin/ddscxxHelloworldSubscriber] Error 1
make[1]: *** [CMakeFiles/Makefile2:231: examples/helloworld/CMakeFiles/ddscxxHelloworldSubscriber.dir/all] Error 2
make: *** [Makefile:152: all] Error 2

```

The same error appears when I try to compile my XTypes tests.

I am not downloading the specific commits that you have indicated to me?

**Comment by reicheratwork** [comment]
The error you get when compiling CDDS-CXX seems to indicate to me that the CDDS-C version you're compiling is the one where a number of things got moved around (i.e. the one that the current PR is not yet in sync with), one of those things being the entity ddsi_default_plist_participant.

When I use the specified commits for CDDS-C and -CXX (using git checkout in stead of git reset to get them addmittedly), CDDS-CXX compiles without issue, as do your examples, and I do not see the issue of not being able to communicate.

This indicates to me, that your CDDS-C is at a commit in front of where 372217a actually is.

**Comment by reicheratwork** [comment]
Could you please tell me the contents of the file ``ddsi_plist.h`` which was installed for CycloneDDS-C, and then especially around line 258?
This file should be (in your case, if I am reading your install etc correctly) located at ``/usr/local/lib/cyclonedds/include/dds/ddsi/ddsi_plist.h``.

**Comment by javiersorianoruiz** [comment]
Yes of course

```
struct ddsi_plist_sample {
  void *blob;
  size_t size;
  nn_parameterid_t keyparam;
};

extern const ddsi_plist_t ddsi_default_plist_participant;     (<------------- this is the line 258 )

/**
 * @brief Initialize global parameter-list parsing indices.
 *
 * These indices are derived from compile-time constant tables.  This only does the work
 * once; ideally it would be done at compile time instead.
 */
void ddsi_plist_init_tables (void);

```

Yes it is located in `/usr/local/lib/cyclonedds/include/dds/ddsi/ddsi_plist.h ` and I have checked that it has got the same content as in my build environment: `/app/cyclonedds-xtypes-testing/cyclonedds/src/core/ddsi/include/dds/ddsi/ddsi_plist.h` (I have used the tool diff to check it)

I have uploaded a copy of this file `ddsi_plist.h` into my XTypes tests repository in the same folder used for the dds logs files (`dds_logs_files` folder in the root directory) in case you want to check it as well.

**Comment by reicheratwork** [comment]
I have a suspicion where things are going wrong on your side and it is somewhere on CycloneDDS-C.
I have made the following branch on my repository which works around this issue: 
[https://github.com/reicheratwork/cyclonedds/commit/e83ca9f8df33eb87004f3719911e7e51680b23bc](https://github.com/reicheratwork/cyclonedds/commit/e83ca9f8df33eb87004f3719911e7e51680b23bc)

If you try this with the previous commit 372217a on my CycloneDDS-CXX repo the linker should now work.

To be fair, this is just a workaround, and not the permanent fix, since I need to work this fix differently with the current state of CycloneDDS-C and then rework the fix in my CXX to be compatible with the current master of CDDS-C. But if this fixes your compile issue, you should see both your publishers communicating with all subscribers.

**Comment by javiersorianoruiz** [comment]
Now it is working with your new branch. Both subscribers receive from both publishers without problems.

I can continue doing more tests to check more XTypes combinations (ej: mutable ...) but I agree with you that all the tests that I am doing, they have to be repeated and verified with the master branch in the official repositories for cyclonedds-cxx and cyclonedds when you finish your fix and merge it.

So, let me know when you do it, and I can repeat XTypes tests for all the implementations (Python, CXX and C)

Thank you very much for your support.

**Comment by reicheratwork** [comment]
@javiersorianoruiz  No problem, and thanks for the effort on your side uncovering this issue.
When I have the CycloneDDS-C fix in and updated PR #331 you will also be referenced :P

**Comment by reicheratwork** [comment]
@javiersorianoruiz  if you could give the following commits a try:

https://github.com/reicheratwork/cyclonedds/commit/bfdddbf75cb1a8af6db45da743035865b4478c4a 
https://github.com/reicheratwork/cyclonedds-cxx/commit/3923b0b15b65561eaf7fa89ab5b6585871b6ba4e

Where the top one is the work waiting to be merged in CDDS-C PR 1503
And the bottom one is the PR for this issue on CDDS-CXX PR 331

**Comment by javiersorianoruiz** [comment]
I have downloaded and installed current master versions of cyclonedds, cyclonedds-python and cyclonedds-cxx and my XTypes test for Cpp and Python are working now properly.

More info about these tests:

https://github.com/javiersorianoruiz/cyclonedds-xtypes-testing

**Comment by javiersorianoruiz** [comment]
I have added to our testing for XTypes, tests for cyclonedds C implementation.

See https://github.com/javiersorianoruiz/cyclonedds-xtypes-testing (subfolder c).

I have used Makefiles written for me to compile these tests (only for the pleasure of remembering how to write Makefiles without cmake).

The C tests are working properly with XTypes, so **I think that this issue #329 can be closed**, because XTypes has been tested for all three implementations (C, CXX and Python) without problems, using the current master of the three projects downloaded and installed yesterday (28-nov-22).
